node/collation-generation/src/lib.rs

```diff
@@ -422,14 +375,277 @@ async fn handle_new_activations<Context>(
+
+	let collation = PreparedCollation {
+		collation,
+		relay_parent,
+		para_id: config.para_id,
+		validation_data,
+		validation_code_hash,
+		n_validators,
+	};
+
+	construct_and_distribute_receipt(
+		collation,
+		config.key.clone(),
+		|msg| ctx.send_message(msg),
+		result_sender,
+		metrics,
+	)
+	.await;
+
+	Ok(())
+}
+
+struct PreparedCollation {
+	collation: Collation,
+	para_id: ParaId,
+	relay_parent: Hash,
+	validation_data: PersistedValidationData,
+	validation_code_hash: ValidationCodeHash,
+	n_validators: usize,
+}
+
+/// Takes a prepared collation, along with its context, and produces a candidate receipt
+/// which is distributed to validators.
+async fn construct_and_distribute_receipt<F: Future<Output = ()>>(
+	collation: PreparedCollation,
+	key: CollatorPair,
+	distribute_collation: impl FnOnce(CollatorProtocolMessage) -> F,
+	result_sender: Option<oneshot::Sender<CollationSecondedSignal>>,
+	metrics: &Metrics,
+) {
+	let PreparedCollation {
+		collation,
+		para_id,
+		relay_parent,
+		validation_data,
+		validation_code_hash,
+		n_validators,
+	} = collation;
+
+	let persisted_validation_data_hash = validation_data.hash();
+	let parent_head_data_hash = validation_data.parent_head.hash();
+
+	// Apply compression to the block data.
+	let pov = {
+		let pov = collation.proof_of_validity.into_compressed();
+		let encoded_size = pov.encoded_size();
+
+		// As long as `POV_BOMB_LIMIT` is at least `max_pov_size`, this ensures
+		// that honest collators never produce a PoV which is uncompressed.
+		//
+		// As such, honest collators never produce an uncompressed PoV which starts with
+		// a compression magic number, which would lead validators to reject the collation.
+		if encoded_size > validation_data.max_pov_size as usize {
+			gum::debug!(
+				target: LOG_TARGET,
+				para_id = %para_id,
+				size = encoded_size,
+				max_size = validation_data.max_pov_size,
+				"PoV exceeded maximum size"
+			);
+
+			return
+		}
+
+		pov
+	};
+
+	let pov_hash = pov.hash();
+
+	let signature_payload = collator_signature_payload(
+		&relay_parent,
+		&para_id,
+		&persisted_validation_data_hash,
+		&pov_hash,
+		&validation_code_hash,
+	);
+
+	let erasure_root = match erasure_root(n_validators, validation_data, pov.clone()) {
+		Ok(erasure_root) => erasure_root,
+		Err(err) => {
+			gum::error!(
+				target: LOG_TARGET,
+				para_id = %para_id,
+				err = ?err,
+				"failed to calculate erasure root",
+			);
+			return
+		},
+	};
+
+	let commitments = CandidateCommitments {
+		upward_messages: collation.upward_messages,
+		horizontal_messages: collation.horizontal_messages,
+		new_validation_code: collation.new_validation_code,
+		head_data: collation.head_data,
+		processed_downward_messages: collation.processed_downward_messages,
+		hrmp_watermark: collation.hrmp_watermark,
+	};
+
+	let ccr = CandidateReceipt {
+		commitments_hash: commitments.hash(),
+		descriptor: CandidateDescriptor {
+			signature: key.sign(&signature_payload),
```

Comment:
Since there is only a single collator key pair in `CollationGenerationConfig` does that mean each relay node talks to at most one collator at a time? Makes sense when the relay node is "in process", but I guess that applies to rpc nodes too.

Reply:
At the moment each relay-chain node can only be one collator on one parachain. Note that this entire subsystem is intended to _run_ on the collators' machine. They send the collations to validators with the `collator-protocol` networking subsystem.

Validators can talk to an arbitrary about of other relay-chain nodes, each of which might be one collator.